src/log.c

```diff
@@ -0,0 +1,361 @@
+    if (msg == NULL)
+        return -1;
+    
+    err = log_send(msg);
+    if (err != EDPVS_OK) {
+        dpvs_log_free(msg);
+        /* log ring is full, need to set limit rate */
+        fprintf(stderr, "log ring is full !\n");
+        log_stats_info[cid].slow = 1;
+        log_stats_info[cid].slow_begin = rte_get_timer_cycles();
+        return -1;
+    }
+    return 0;
+}
+
+int dpvs_log(uint32_t level, uint32_t logtype, const char *func, int line, const char *format, ...)
+{
+    va_list ap;
+    lcoreid_t cid;
+    char log_buf[DPVS_LOG_MAX_LINE_LEN];
+    int len = 0;
+    int off = g_dpvs_log_time_off;
+
+    if (level > rte_logs.level)
+        return -1;
+
+    va_start(ap, format);
+
+    do { 
+        if (!g_dpvs_log_async_mode || !g_dpvs_log_core || !g_dpvs_log_thread_ready) {
+            rte_vlog(level, logtype, format, ap);
+            break;
+        }
+
+        /* async log is not used for ctrl message */
+        if (logtype != RTE_LOGTYPE_USER1) {
+            rte_vlog(level, logtype, format, ap);
+            break;
+        }
+
+        cid = rte_lcore_id();
+
+        if (log_stats_info[cid].slow) {
+            /* set log limit rate to 5 sec and keep for 10 mins */
+            if (rte_get_timer_cycles() - log_stats_info[cid].slow_begin > LOG_SLOW_INTERNAL_TIME*rte_get_timer_hz()) {
+                log_stats_info[cid].slow = 0;
+            }
+            if ((rte_get_timer_cycles() - log_stats_info[cid].log_begin) < log_internal*rte_get_timer_hz()) {
+                log_stats_info[cid].missed++;
+                break;
+            }
+            /* just output func and line if log is too fast */
+            len = snprintf(log_buf+off, sizeof(log_buf)-off, "%s:%d\n", func, line);
+            dpvs_async_log(level, logtype, cid, log_buf, len, off);
+            break;
+        }
+
+        len = vsnprintf(log_buf+off, sizeof(log_buf)-off, format, ap);                
+        dpvs_async_log(level, logtype, cid, log_buf, len, off);
+    }while(0);
+    
+    va_end(ap);
+    return 0;
+}
+
+static int log_buf_flush(FILE *f)
+{
+    if (f == NULL ) {
+        w_buf.buf[w_buf.pos] = '\0';
+        syslog(w_buf.level, "%s", w_buf.buf);
+    } else {
+        fwrite(w_buf.buf, w_buf.pos, sizeof(w_buf.buf[0]), f);
+        fflush(f);
+    }
+    w_buf.pos = 0;
+    return 0;
+}
+
+static int log_buf_timeout_flush(FILE *f, int timeout)
+{
+    uint64_t now;
+    
+    now = rte_get_timer_cycles();
+    
+    if (w_buf.pos && ((now - w_buf.time) >= timeout*rte_get_timer_hz())) {
+        log_buf_flush(f);
+    }
+    return 0;
+}
+
+static int log_slave_process(void)
+{
+    struct dpvs_log *msg_log;
+    int ret = EDPVS_OK;
+    FILE *f = rte_logs.file;
+
+    /* dequeue LOG from ring, no lock for ring and w_buf */
+    while (0 == rte_ring_dequeue(log_ring, (void **)&msg_log)) { 
+        if (msg_log->log_len > LOG_BUF_MAX_LEN) {
+            dpvs_log_free(msg_log);
+            continue;
```

Comment:
Sliently drop the log message here? Can we not enqueue the oversized log message and return failure when sent?